superchainERC20 (#986)
ethereum-optimism/docs: pages/stack/protocol/interop/superchain-erc20.mdx

---
title: SuperchainERC20 Token Standard
lang: en-US
description: Learn basic details about the SuperchainERC20 token standard.
---

import { Callout } from 'nextra/components'

# SuperchainERC20 Token Standard

<Callout>
  Interop is currently in active development and not yet ready for production use. The information provided here may change. Check back regularly for the most up-to-date information.
</Callout>

> `SuperchainERC20` is a token standard designed to enable asset interoperability in the Superchain ecosystem. 

Reviewer: tbh you could get away with just saying 

`SuperchainERC20` is a token standard designed to enable asset interoperability in the Superchain.

Not sure what we're adding with (It serves as...) and (This standard is crucial for...)

Reviewer: Consider exapnding on what asset interoperability means.....it allows for tokens to securely move across chains without asset wrapping or liquidity pools for maximal capital efficiency - unifying liquidity and simplifying user experience.
 
 Should we talk about how it works?
 
This standard work facilitates secure token transfers between chains in the Superchain networks via native burning and minting.
 
 Initating message where Token is burned on the source chain: a user initiates a transfer of token from one blockchain to another, and specifies the recipient wallet address on the destination chain. A specified amount of token is burned on the source chain.
 
 Executing message where token is minted on the destination chain. The specified amount of token is minted on the destination chain and sent to the recipient wallet address.

> It serves as the foundational token deployment pattern within the Superchain, including `GovToken`, `WETH` (Wrapped Ether), migrated `OptimismMintableERC20Tokens`, and newly minted native `SuperchainERC20`. 

Author: removed and replaced with lang above

> `SuperchainERC20` is a token standard specifically designed for the Superchain ecosystem. It serves as the foundation for interoperable tokens within the Superchain, including `GovToken`, `WETH` (Wrapped Ether), and migrated `OptimismMintableERC20Tokens`. This standard is crucial for enabling a more interconnected and efficient token ecosystem within the Superchain, balancing the needs for interoperability, security, and simplified governance.

Reviewer 2: > `SuperchainERC20` is a token standard specifically designed for the Superchain ecosystem.
This should probably lean more in the Ethereum crosschain standardization. Maybe something like:

"`SuperchainERC20` is a token standard designed to enable asset interoperability in the Superchain ecosystem. It serves as the foundational token deployment pattern within the Superchain, including `GovToken`, `WETH` (Wrapped Ether), migrated `OptimismMintableERC20Tokens`, and newly minted native `SuperchainERC20`. This standard is crucial for enabling a more interconnected and efficient token ecosystem within the Superchain, balancing the needs for interoperability, security, and simplified developer experience."

Author: love it! updated!

## Features and benefits

*   **Interoperability**: Enables seamless token transfers between different chains within the Superchain ecosystem.

> *   **Built-in bridging**: The bridging interface exists on the token itself, eliminating the need for a central bridge predeploy (e.g., `bridgeERC20`, `bridgeERC20To`).

Reviewer 2: Not sure I follow this one?

Author: no prob, removing it 😆

> *   **Consistency**: Provides a unified standard for tokens across all Superchain-compatible networks.

Author: updated the lang. can you send the link over pls?

> *   **Governance minimized**: Designed to keep tokens within the Superchain ecosystem with minimal governance intervention.

Author: yep, removed! in both places!

*   **Permissionless propagation**: Easily clone an existing token contract to a new OP Stack chain using `create2` without requiring the original owner, which enables movement of assets to the new chain once Interop goes live. Importantly, permissionless propagation retains the integrity of the original owner on the contract and preserves security but proliferates the contract's availability to new chains.

> *   **Ethereum-aligned**: Intentionally designed to be supported across the EVM ecosystem (RIP coming soon).

Reviewer: Prob a better way to say this. 

**Ethereum-aligned**: Intentionally designed to be generic and supported as an Ethereum-wide standard (RIP coming soon).

Author: updated

## Major components

*   **Token Contract**: implements the `SuperchainERC20` standard with bridging functionality.
*   **Factory Predeploy**: uses a `create3`-based factory for deploying `SuperchainERC20` tokens consistently across chains.

> *   **Bridging Functions**: using methods like `bridgeERC20` and `bridgeERC20To` for cross-chain transfers.

Author: great catch! thanks!

*   **Minting Role**: modifies the definition to be minting role-based rather than interface-based.

## Comparison to other standards

> `SuperchainERC20` differs from other token standards in its focus and implementation:

Author: yay! thank you!

`SuperchainERC20` differs from other token standards in its focus and implementation:

*   `SuperchainERC20` has minimal differentiation from a standard ERC20 deployment, only requiring a minimal crosschain mint/burn interface, which aims to be a common pattern for the EVM ecosystem (RIP coming soon).

> *   `SuperchainERC20` shares trust assumptions across all chains in the superchain, such that custom assumptions around security and latency are not required to account for when executing transfers.

Reviewer: S in Superchain should be uppercase

<Callout type="info">
  Projects moving from other token standards may need to adapt to the `SuperchainERC20` specification.
</Callout>

## Implementation details

Application developers must complete two tasks to enable `SuperchainERC20`. This advance setup ensures that `SuperchainERC20` goes live when the interop upgrade happens and your token becomes interoperable.

1.  Deployment at same address on each chain using `create2` function

> 2.  Permission `crosschainMint` and `crosschainBurn` to the Superchain bridge address

Author: agreed, updated!

> Application developers must complete two tasks to enable `SuperchainERC20`. This advance setup ensures that `SuperchainERC20` goes live when the Interop upgrade happens and tokens becomes interoperable.

Reviewer: This is a weird paragraph. 

We should be trying to say:
-App devs must do two things to make their tokens SuperchainERC20 compatible
-Doing this setup today will ensure that their tokens can benefit from Interop once the Interop upgrade happens

> For now, application developers should view `SuperchainERC20`as standard ERC20 tokens with additional built-in functions that will be enabled once Interop goes live.

Reviewer: built-in functions that allow cross-chain asset movement that will be enabled once Interop goes live.

<Callout type="warning">

>   `SuperchainERC20` token must give access to the address where the future Interop bridge will live.

Reviewer: In order to enable asset interoperability, `SuperchainERC20` token must give access to the address where the future`SuperchainERC20Bridge` will live.

---
title: SuperchainERC20 Token Standard
lang: en-US
description: Learn basic details about the SuperchainERC20 token standard.
---

import { Callout } from 'nextra/components'

# SuperchainERC20 Token Standard

<Callout>
  Interop is currently in active development and not yet ready for production use. The information provided here may change. Check back regularly for the most up-to-date information.
</Callout>

`SuperchainERC20` is a token standard designed to enable asset interoperability in the Superchain. 
Asset interoperability allows for tokens to securely move across chains without asset wrapping or liquidity pools for maximal capital efficiency, thus unifying liquidity and simplifying the user experience.

Additional features: 
*   **Simplified deployments**: Provides a consistent, unified standard for tokens across all Superchain-compatible networks and a common crosschain interface for the EVM ecosystem at large.

> *   **Permissionless propagation**: Easily clone an existing token contract to a new OP Stack chain using `create2` without requiring the original owner, which enables movement of assets to the new chain once Interop goes live. Importantly, permissionless propagation retains the integrity of the original owner on the contract and preserves security but proliferates the contract's availability to new chains.

Reviewer 3: ```suggestion
*   **Permissionless propagation**: Easily clone an existing token contract to a new OP Stack chain using `create2` without requiring the original deployer, which enables movement of assets to the new chain once Interop goes live. Importantly, permissionless propagation retains the integrity of the original owner on the contract and preserves security but proliferates the contract's availability to new chains.
```

*   **Ethereum-aligned**: Intentionally designed to be generic and supported as an Ethereum-wide standard (RIP coming soon).

## How it works

`SuperchainERC20` token standard facilitates secure token transfers between chains in the Superchain networks via native burning and minting.

*   Initiating message where token is **burned** on the source chain. A user initiates a transfer of token from one blockchain to another and specifies the recipient wallet address on the destination chain. A specified amount of token is burned on the source chain.

> *   Executing message where token is **minted** on the destination chain. The specified amount of token is minted on the destination chain and sent to the recipient wallet address.

Reviewer 3: ```suggestion
*   Executing message where token is **minted** on the destination chain. The specified amount of token is minted on the destination chain directly to the recipient wallet address.
```

## Major components

*   **Token Contract**: implements the `SuperchainERC20` standard with bridging functionality.
*   **Factory Predeploy**: uses a `create2`-based factory for deploying `SuperchainERC20` tokens consistently across chains.
*   **Bridging Functions**: using methods like `sendERC20` and `relayERC20` for cross-chain transfers.

> *   **Minting Role**: modifies the definition to be minting role-based rather than interface-based.

Reviewer: should be removed

> *   **Minting Role**: modifies the definition to be minting role-based rather than interface-based.

Reviewer 3: ```suggestion
```

## Comparison to other standards

`SuperchainERC20` differs from other token standards in its focus and implementation:

*   `SuperchainERC20` has minimal differentiation from a standard ERC20 deployment, only requiring a minimal crosschain mint/burn interface, which aims to be a common pattern for the EVM ecosystem (RIP coming soon).
*   `SuperchainERC20` shares trust assumptions across all chains in the Superchain, such that custom assumptions around security and latency are not required to account for when executing transfers.

<Callout type="info">
  Projects moving from other token standards may need to adapt to the `SuperchainERC20` specification.
</Callout>

## Implementation details

Application developers must do two things to make their tokens `SuperchainERC20` compatible. Doing this setup now ensures that tokens can benefit from Interop once the Interop upgrade happens.

1.  Permission only `SuperchainERC20Bridge` to call `crosschainMint` and `crosschainBurn`.
2.  Deployment at same address on every chain in the Superchain using `create2` function.

> For now, application developers should view `SuperchainERC20`as standard ERC20 tokens with additional built-in functions that allow cross-chain asset movement that will be enabled once Interop goes live.

Reviewer 3: ```suggestion
For now, application developers should view `SuperchainERC20` as standard ERC20 tokens with additional built-in functions that allow cross-chain asset movement that will be enabled once Interop goes live.
```

<Callout type="warning">
  To enable asset interoperability, `SuperchainERC20` token must give access to the address where the future `SuperchainERC20Bridge` will live.
</Callout>

## Next steps

> *   Join the discussion! `SuperchainERC20` implementation is being developed in partnership with Wonderland, with discussions taking place in the [#interop-evm](https://discord.gg/3zvcEFj6Nu) Discord channel.

Reviewer 3: should this invite link be public?

Author: @skeletor-spaceman ooh i'm not sure, would you like me to remove it for now? [just the link]

Author: cc @zainbacchus @pharger

Reviewer 3: if we remove please also let's disable it from discord too, since it will remain on GH history :)

Author: removing based on convos

---
title: SuperchainERC20 Token Standard
lang: en-US
description: Learn basic details about the SuperchainERC20 token standard.
---

import { Callout } from 'nextra/components'

# SuperchainERC20 Token Standard

<Callout>
  Interop is currently in active development and not yet ready for production use. The information provided here may change. Check back regularly for the most up-to-date information.
</Callout>

`SuperchainERC20` is a token standard designed to enable asset interoperability in the Superchain.
Asset interoperability allows for tokens to securely move across chains without asset wrapping or liquidity pools for maximal capital efficiency, thus unifying liquidity and simplifying the user experience.

Additional features:

*   **Simplified deployments**: Provides a consistent, unified standard for tokens across all Superchain-compatible networks and a common crosschain interface for the EVM ecosystem at large.

> *   **Permissionless propagation**: Easily clone an existing token contract to a new OP Stack chain using `create2` without requiring the original owner, which enables movement of assets to the new chain once Interop goes live. Importantly, permissionless propagation retains the integrity of the original owner on the contract and preserves security but proliferates the contract's availability to new chains.

Reviewer 4: The standard itself does not enforce any deployment method. A token can be deployed using CREATE or an EOA as deployer, so posterior deployments might result in a different address.

> *   **Ethereum-aligned**: Intentionally designed to be generic and supported as an Ethereum-wide standard (RIP coming soon).

Reviewer 4: nit: RIP is probably gonna be an ERC

```mermaid
sequenceDiagram
		box rgba(255, 4, 32, 0.1) ChainA
		  participant User-op
		  participant SuperchainERC20-op
	    participant SuperchainERC20Bridge-op
		end
		box rgba(248, 61, 213, 0.1) ChainB
			participant SuperchainERC20Bridge-uni
	    participant SuperchainERC20-uni
	    participant User-uni
		end
  
    
    User-op->>SuperchainERC20-op: Initiate token transfer
    SuperchainERC20-op->>SuperchainERC20Bridge-op: Bridge to Uni
    SuperchainERC20Bridge-op->>SuperchainERC20-op: Burn tokens
    SuperchainERC20Bridge-op-->>SuperchainERC20Bridge-op: Emit cross-chain event
    SuperchainERC20Bridge-uni-->>SuperchainERC20Bridge-uni: Validates message
    SuperchainERC20Bridge-uni-->>SuperchainERC20-uni: Mint tokens
    SuperchainERC20-uni->>User-uni: User receives tokens
```

This diagram illustrates the process where tokens are burned on the source chain and minted on the destination chain, enabling seamless cross-chain transfers without the need for asset wrapping or liquidity pools.

## Major components

*   **Token Contract**: implements the `SuperchainERC20` standard with bridging functionality.

> *   **Factory Predeploy**: uses a `create2`-based factory for deploying `SuperchainERC20` tokens consistently across chains.

Reviewer 4: This is not enforced in the token standard. The `OptimismSuperchainERC20` (specific implementation of `SuperchainERC20` that can be converted back and forth with `OptimismMintableERC20`) will use an official factory (the `OptimismSuperchainERC20Factory` predeploy), but its not a requisite for every token. 
Every token does need same address across different chains to be processed by the bridge. We recommend deployments happen with a standarized method, such as the `Create2Deployer`, but again, the standard is as neutral as possible on this point.

> *   **Bridging Functions**: using methods like `sendERC20` and `relayERC20` for cross-chain transfers.

Reviewer 4: you can mention these live in the Bridge contract, and not in the token itself

## Comparison to other standards

`SuperchainERC20` differs from other token standards in its focus and implementation:

*   `SuperchainERC20` has minimal differentiation from a standard ERC20 deployment, only requiring a minimal crosschain mint/burn interface, which aims to be a common pattern for the EVM ecosystem (RIP coming soon).
*   `SuperchainERC20` shares trust assumptions across all chains in the Superchain, such that custom assumptions around security and latency are not required to account for when executing transfers.

<Callout type="info">
  Projects moving from other token standards may need to adapt to the `SuperchainERC20` specification.
</Callout>

## Implementation details

Application developers must do two things to make their tokens `SuperchainERC20` compatible. Doing this setup now ensures that tokens can benefit from Interop once the Interop upgrade happens.

1.  Permission only `SuperchainERC20Bridge` to call `crosschainMint` and `crosschainBurn`.

> 2.  Deployment at same address on every chain in the Superchain using `create2` function.

Reviewer 4: nit: did you mean the `Create2Deployer` preinstall here? or the `CREATE2` opcode?

*   Initiating message where token is **burned** on the source chain. A user initiates a transfer of token from one blockchain to another and specifies the recipient wallet address on the destination chain. A specified amount of token is burned on the source chain.
*   Executing message where token is **minted** on the destination chain. The specified amount of token is minted on the destination chain and sent to the recipient wallet address.

```mermaid
sequenceDiagram
    participant User
    participant SourceChain
    participant SuperchainERC20Bridge
    participant DestinationChain

    User->>SourceChain: Initiate token transfer
    SourceChain->>SourceChain: Burn tokens
    SourceChain-->>SuperchainERC20Bridge: Emit cross-chain event
    SuperchainERC20Bridge-->>DestinationChain: Relay event
    DestinationChain->>DestinationChain: Mint tokens
    DestinationChain->>User: User receives tokens

> ```

Reviewer 3: https://specs.optimism.io/interop/token-bridging.html?highlight=token-bridg#diagram
you can use this on the example, since it's a bit more precise on the flow

Author: @skeletor-spaceman @pharger -- we were thinking the simplified flow maybe worked better in docs since this is a high-level explainer (used by broader audience). Does that check out?

Reviewer 3: it does, but the current `mermaid` flow will, imho, confuse more than a slightly more complex one.
